frontendGraph/src/main/java/com/virtuslab/gitmachete/frontend/graph/nodes/BaseGraphNode.java

```diff
@@ -0,0 +1,72 @@
+package com.virtuslab.gitmachete.frontend.graph.nodes;
+
+import lombok.Getter;
+import lombok.ToString;
+import org.checkerframework.checker.index.qual.GTENegativeOne;
+import org.checkerframework.checker.index.qual.NonNegative;
+import org.checkerframework.checker.index.qual.Positive;
+import org.checkerframework.checker.nullness.qual.MonotonicNonNull;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+import com.virtuslab.gitmachete.frontend.graph.coloring.GraphEdgeColor;
+
+@Getter
+@ToString
+public abstract class BaseGraphNode implements IGraphNode {
+
+  @Nullable
+  private final GraphEdgeColor graphEdgeColor;
+
+  @GTENegativeOne
+  private final int prevSiblingNodeIndex;
+
+  @Positive
+  @MonotonicNonNull
+  private Integer nextSiblingNodeIndex = null;
+
+  @NonNegative
+  private final int indentLevel;
+
+  protected BaseGraphNode(@Nullable GraphEdgeColor graphEdgeColor,
+      @GTENegativeOne int prevSiblingNodeIndex,
+      @Positive int nextSiblingNodeIndex,
+      @NonNegative int indentLevel) {
+    this.graphEdgeColor = graphEdgeColor;
+    this.prevSiblingNodeIndex = prevSiblingNodeIndex;
+    this.nextSiblingNodeIndex = nextSiblingNodeIndex;
+    this.indentLevel = indentLevel;
+  }
+
+  protected BaseGraphNode(@Nullable GraphEdgeColor graphEdgeColor,
+      @GTENegativeOne int prevSiblingNodeIndex,
+      @NonNegative int indentLevel) {
+    this.graphEdgeColor = graphEdgeColor;
+    this.prevSiblingNodeIndex = prevSiblingNodeIndex;
+    this.indentLevel = indentLevel;
+  }
+
+  @Override
+  @NonNegative
+  public int getIndentLevel() {
+    return indentLevel;
+  }
+
+  @Override
+  public void setNextSiblingNodeIndex(@Positive int i) {
+    assert nextSiblingNodeIndex == null : "nextSiblingNodeIndex has already been set";
+    nextSiblingNodeIndex = i;
+  }
+
+  @Override
+  public final boolean equals(@Nullable Object other) {
+    return this == other;
+  }
```

Comment:
This is the default implementation of equals, so I think this can be omitted unless it's for clarity

Reply:
The idea is it should be provided as `final` so that nobody can override it subclasses (note that https://checkerframework.org/api/org/checkerframework/checker/interning/qual/UsesObjectEquals.html can be used for this purpose instead... but so far the `InterningChecker` isn't wired up)